File: Arrays/Algorithms/Sliding_Window/Frequency_Based/count_substrings_with_constraints.py
def count_substrings_with_k_distinct(s, k):
	from collections import defaultdict
	freq = defaultdict(int)
	l = res = 0
	for r in range(len(s)):
		freq[s[r]] += 1
		while len(freq) > k:
			freq[s[l]] -= 1
			if freq[s[l]] == 0:
				del freq[s[l]]
			l += 1
		res += r - l + 1
	return res

if __name__ == "__main__":
	s = input().strip()
	k = int(input())
	print(count_substrings_with_k_distinct(s, k))

File: Arrays/Algorithms/Sliding_Window/Frequency_Based/count_substrings_with_constraints.cpp
#include <bits/stdc++.h>
using namespace std;

int countSubstringsWithKDistinct(const string& s, int k) {
	unordered_map<char, int> freq;
	int l = 0, res = 0;
	for (int r = 0; r < s.size(); r++) {
		freq[s[r]]++;
		while (freq.size() > k) {
			freq[s[l]]--;
			if (freq[s[l]] == 0) freq.erase(s[l]);
			l++;
		}
		res += r - l + 1;
	}
	return res;
}

int main() {
	string s;
	int k;
	cin >> s >> k;
	cout << countSubstringsWithKDistinct(s, k) << endl;
	return 0;
}
